Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp

```diff
@@ -276,6 +279,48 @@ void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL
     m_context->simulateEventForTesting(event);
 }
 
+void RemoteGraphicsContextGL::getBufferSubDataInline(uint32_t target, uint64_t offset, size_t dataSize, CompletionHandler<void(std::span<const uint8_t>)>&& completionHandler)
+{
+    assertIsCurrent(workQueue());
+    static constexpr size_t getBufferSubDataInlineSizeLimit = 64 * KB; // NOTE: when changing, change the value in RemoteGraphicsContextGLProxy too.
+
+    MESSAGE_CHECK(dataSize <= getBufferSubDataInlineSizeLimit);
+
+    if (!dataSize) {
+        completionHandler({ });
+        return;
+    }
+    MallocPtr<uint8_t> bufferStore;
+    std::span<uint8_t> bufferData;
+    bufferStore = MallocPtr<uint8_t>::tryMalloc(dataSize);
+    if (bufferStore) {
+        bufferData = { bufferStore.get(), dataSize };
+        if (!m_context->getBufferSubDataWithStatus(target, offset, bufferData))
+            bufferData = { };
+    } else
+        m_context->addError(GCGLErrorCode::OutOfMemory);
+
+    completionHandler(bufferData);
+}
+
+void RemoteGraphicsContextGL::getBufferSubDataSharedMemory(uint32_t target, uint64_t offset, size_t dataSize, WebCore::SharedMemory::Handle handle, CompletionHandler<void(bool)>&& completionHandler)
+{
+    assertIsCurrent(workQueue());
+    bool validBufferData  = false;
+
+    static constexpr size_t readPixelsSharedMemorySizeLimit = 100 * MB; // NOTE: when changing, change the value in RemoteGraphicsContextGLProxy too.
+
+    MESSAGE_CHECK(dataSize <= readPixelsSharedMemorySizeLimit);
+
+    handle.setOwnershipOfMemory(m_sharedResourceCache->resourceOwner(), WebKit::MemoryLedger::Default);
+    if (auto buffer = SharedMemory::map(WTFMove(handle), SharedMemory::Protection::ReadWrite))
```

Comment:
this is still lacking the size check.
E.g. if caller sends 4kb memory and 50mbs dataSize, it's not going well.